### PR TITLE
`ruff server` now supports commands for auto-fixing, organizing imports, and formatting

### PR DESCRIPTION
**Pull request by snowsignal**
## Summary

This builds off of the work in https://github.com/astral-sh/ruff/pull/10652 to implement a command executor, backwards compatible with the commands from the previous LSP (`ruff.applyAutofix`, `ruff.applyFormat` and `ruff.applyOrganizeImports`). 

This involved a lot of refactoring and tweaks to the code action resolution code - the most notable change is that workspace edits are specified in a slightly different way, using the more general `changes` field instead of the `document_changes` field (which isn't supported on all LSP clients). Additionally, the API for synchronous request handlers has been updated to include access to the `Requester`, which we use to send a `workspace/applyEdit` request to the client.

## Test Plan

https://github.com/astral-sh/ruff/assets/19577865/7932e30f-d944-4e35-b828-1d81aa56c087

### CODE REVIEW

**Comment by snowsignal** [comment]
(This is in draft because I'd like to hold off reviews until the downstream PRs are merged in, and also because I still need to make a test plan video for this)

**Comment by MichaReiser** [comment]
Sorry for jumping in on a Draft PR. Before publishing, would you mind explaining what happens for editors not supporting the `changes` field?

> the most notable change is that workspace edits are specified in a slightly different way, using the more general changes field instead of the document_changes field (which isn't supported on all LSP clients).

**Comment by snowsignal** [comment]
> Sorry for jumping in on a Draft PR. Before publishing, would you mind explaining what happens for editors not supporting the `changes` field?

It's the `document_changes` field that isn't supported on all clients. `changes` should have universal support.

From the [LSP specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspaceEdit):
```
If a client neither supports `documentChanges` nor `workspace.workspaceEdit.resourceOperations`,
then only plain `TextEdit`s using the `changes` property are supported.
```

**Comment by MichaReiser** [comment]
Would you mind explaining the reasoning behind this

> This involved a lot of refactoring and tweaks to the code action resolution code - the most notable change is that workspace edits are specified in a slightly different way, using the more general changes field instead of the document_changes field (which isn't supported on all LSP clients). 

From the LSP specification

> A workspace edit represents changes to many resources managed in the workspace. The edit should either provide changes or documentChanges. If the client can handle versioned document edits and if documentChanges are present, the latter are preferred over changes.

So it seems we should be using documentChanges when we can.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (MichaReiser) [inline crates/ruff_server/src/server/api/requests/execute_command.rs:46]: Nit: We could incline `args_as_text_documents` to avoid the vec allocation and without increasing the complexity (I almost find it easier to read because I don't need to jump to `args_as_text_documents` to understand what's happening)
```suggestion
        for argument in params.arguments {
            let document = serde_json::from_value(argument).with_failure_code(ErrorCode::InvalidParams)?;
            let snapshot = session
                .take_snapshot(&document.uri)
```

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (MichaReiser) [inline crates/ruff_server/src/server/api/requests/execute_command.rs:91]: What's the motivation for testing this at the very end? Should we instead not promote the server capability (or commands) if the client doesn't support `workspace/applyEdit`?

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (MichaReiser) [inline crates/ruff_server/src/server/api/requests/execute_command.rs:88]: Hmm, this could get interesting once we have fixes that can "fix" across file because it is then no longer guaranteed that the changes are not overlapping. But that's something to worry about at another day. 

What we could consider doing is to move this inside the loop?

**Review by MichaReiser — COMMENTED** [review]
Thanks for the extensive test plan. 

This looks good code wise. I'm not sure about migrating from `documentChanges` to `changes`. I would like to understand the motivation/reason for this better.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (dhruvmanila) [inline crates/ruff_server/src/server/api/requests/execute_command.rs:13]: nit: It's common to at least have the `Debug` trait derive on types.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (dhruvmanila) [inline crates/ruff_server/src/server/api/requests/execute_command.rs:39]: nit: we could possibly implement `FromStr` for `Command` and then use `?` operator to avoid creating an empty `anyhow` error. I'm not sure how will `with_failure_code` comes from which I believe you're aware of.

```rs
impl FromStr for Command {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match command {
            "ruff.applyAutofix" => Self::FixAll,
            "ruff.applyFormat" => Self::Format,
            "ruff.applyOrganizeImports" => Self::OrganizeImports,
            _ => return Err(()),
        })
    }
}
```

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (dhruvmanila) [inline crates/ruff_server/src/server/api/requests/execute_command.rs:43]: Can you confirm whether `changes` support having multiple documents? I can't exactly figure out reading the spec but I presume it does. But, if it doesn't then this will be problematic for Jupyter Notebooks where each cell is its own text document so the edits are tied to multiple documents.

**Review by dhruvmanila — COMMENTED** [review]
Thank you! I'll test this now in Neovim.

I'd also like to understand the migration from `document_changes` to `changes`. I've pointed out the reason for that.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (MichaReiser) [reply crates/ruff_server/src/server/api/requests/execute_command.rs:39]: Good catch. I think we should always avoid anyhow error.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (snowsignal) [reply crates/ruff_server/src/server/api/requests/execute_command.rs:88]: Multi-file fixes are going to require a lot of refactoring regardless, but yes, this is definitely something we'll need to solve eventually. For now though, it's a future concern.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (snowsignal) [reply crates/ruff_server/src/server/api/requests/execute_command.rs:91]: Good idea, I'll move this up.

**Comment by snowsignal** [comment]
The main reason I switched from `documentChanges` to `changes` was over concerns about supporting older versions. I think what I'll do is refactor our code to support `documentChanges` but use `changes` as a fallback.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (snowsignal) [reply crates/ruff_server/src/server/api/requests/execute_command.rs:43]: `changes` does enable multi-document edits, though none of the commands will modify multiple files. I think the LSP spec allows treating notebook files as a single document, so we could still have a single `changes` entry that covers multiple cells in a notebook.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (snowsignal) [reply crates/ruff_server/src/server/api/requests/execute_command.rs:39]: We'd still need to define an error type for `FromStr` - I think using anyhow as an error makes sense here because it lets us pass an arbitrary error message. The message being empty here was just an oversight.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (charliermarsh) [inline crates/ruff_server/src/server/api/requests/execute_command.rs]: I would suggest perhaps passing in whatever's returned by `session.resolved_client_capabilities()` to reduce the risk of passing in a wrong boolean.

File: crates/ruff_server/src/server/api/requests/code_action.rs
Comment (charliermarsh) [inline crates/ruff_server/src/server/api/requests/code_action.rs:63]: Can we use `document_changes` here?

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (charliermarsh) [inline crates/ruff_server/src/server/api/requests/execute_command.rs:56]: So is it guaranteed that documents aren't repeated here?

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (charliermarsh) [inline crates/ruff_server/src/server/api/requests/execute_command.rs:154]: I naively would've assumed that this would _replace_ or _augment_ the existing edits, rather than erroring. I might suggest making it a replacement (`set_edits_for_document`)?

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (charliermarsh) [inline crates/ruff_server/src/server/api/requests/execute_command.rs:142]: Do you mind adding a rustdoc for this one?

**Review by charliermarsh — APPROVED** [review]
I would be pro-using `document_changes` everywhere, since that's what we do in the existing LSP and it's never been an issue. My _guess_ is that it's pretty rare for LSP's _not_ to support that (or, at least, that they must be very old), and I'm just not really clear on the implications of using `changes` in various places.

(I like that you made it a fallback in some places, but it looks like `quick_fix` at least still uses `changes` always.)

So I'd vote to continue using `document_changes` everywhere until it proves to be a problem. But it doesn't need to block this PR. You could also make that change (revert the use of `changes`) and merge without a re-review.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (snowsignal) [reply crates/ruff_server/src/server/api/requests/execute_command.rs:56]: We enforce it indirectly by preventing re-setting edits on a document, but we should probably enforce it here too.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (charliermarsh) [reply crates/ruff_server/src/server/api/requests/execute_command.rs:154]: It seems like one of the goals of `document_changes` is that it's _ok_ to repeat a URI as long as the version differs? So perhaps this check should be on URI and version?

**Comment by snowsignal** [comment]
You're probably right that `changes` could be supported without issue. My general policy with features tied to client capabilities is that we _should_ have a fall-back in case that capability doesn't exist, but I should reconsider whether that makes sense going forward. I'm going to err on the side of backwards compatibility for now.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (snowsignal) [reply crates/ruff_server/src/server/api/requests/execute_command.rs]: Still, I don't think it makes sense to apply document edits to the same document at different points in time simultaneously.

File: crates/ruff_server/src/server/api/requests/execute_command.rs
Comment (dhruvmanila) [reply crates/ruff_server/src/server/api/requests/execute_command.rs]: > I think the LSP spec allows treating notebook files as a single document, so we could still have a single `changes` entry that covers multiple cells in a notebook.

Yes, but it's not about a Notebook document here. How would the client know which _cell_ does this edit belong to? This is only known through the text document URI because each cell is represented using text document. Refer to [this loop](https://github.com/astral-sh/ruff-lsp/blob/187d7790be0783b9ac41ce025a724cf389bf575c/ruff_lsp/server.py#L1352-L1367) which does that in `ruff-lsp`.

**Pull request merged**
